src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -408,87 +436,38 @@ private int FillBuffer(int minSize)
                         throw new IOException(SR.net_io_eof);
                     }
 
-                    return 0;
+                    return new ValueTask<int>(0);
                 }
 
                 _internalBufferCount += bytes;
             } while (_internalBufferCount < minSize);
 
-            return minSize;
-        }
+            return new ValueTask<int>(minSize);
 
-        private static void CompleteFromCompletedTask(Task<int> task, AsyncProtocolRequest asyncRequest)
-        {
-            Debug.Assert(task.IsCompleted);
-            if (task.IsCompletedSuccessfully)
+            async Task<int> InternalFillBufferAsync(TReadAdapter adap, Task<int> task, int min, int initial)
             {
-                asyncRequest.CompleteRequest(task.Result);
-            }
-            else if (task.IsFaulted)
-            {
-                asyncRequest.CompleteUserWithError(task.Exception.InnerException);
-            }
-            else
-            {
-                asyncRequest.CompleteUserWithError(new OperationCanceledException());
-            }
-        }
-
-        // Returns true if pending, false if completed
-        private static bool TaskToAsyncProtocolRequest(Task<int> task, AsyncProtocolRequest asyncRequest, AsyncProtocolCallback asyncCallback, out int result)
-        {
-            // Parameters other than asyncCallback are not used
-            asyncRequest.SetNextRequest(null, 0, 0, asyncCallback);
-
-            if (task.IsCompleted)
-            {
-                CompleteFromCompletedTask(task, asyncRequest);
-            }
-            else
-            {
-                task.ContinueWith((t, s) => CompleteFromCompletedTask(t, (AsyncProtocolRequest)s),
-                    asyncRequest,
-                    CancellationToken.None,
-                    TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.DenyChildAttach,
-                    TaskScheduler.Default);
-            }
-
-            if (asyncRequest.MustCompleteSynchronously)
-            {
-                result = asyncRequest.Result;
-                return false;
-            }
+                while (true)
+                {
+                    int b = await task.ConfigureAwait(false);
```

Comment:
Is this supposed to be outside of the loop?  Otherwise on subsequent iterations we're doing an await again on a task we already know is completed, and we're overwriting `b` with what's presumably the incorrect value.  If I'm correct about that, are we missing a test?

Reply:
the bottom should have been 

`task = ReadAsync(......`

Reply:
I wonder if this was the cause of some of the repeatable outloop failures... lets see